activerecord/test/cases/database_configurations/resolver_test.rb

```diff
@@ -51,7 +51,7 @@ def test_url_sub_key
         end
 
         def test_url_sub_key_merges_correctly
-          hash = { "url" => "abstract://foo?encoding=utf8&", "adapter" => "sqlite3", "host" => "bar", "pool" => "3" }
+          hash = { "url" => "abstract://foo?encoding=utf8&", "host" => "bar", "pool" => "3" }
```

Comment:
If this test needs to change in order to get the new test passing, I think it needs another solution. Otherwise this could break existing  behavior.

Reply:
Good point. I just remember that the expected behavior is that `url` has precedence over any configuration set in the config. This is changing the behavior to only make adapter special and I don't think we should change that behavior.

To me it seems the correct fix for `#41137` is that the adapter rename the `url` config that it expects to have a different behavior to something else that is not used by the framework.

Reply:
Yea I agree, but thought it was weird to intermingle the adapter setting with url. And things like:

```yaml
default: &default
  url: postgresql://host/foo

production:
  <<: *default
  adapter: sqlite3
  database: bar
```

would really throw someone for a loop :)

But given the preference for URL as highest maybe we could check to make sure it's a valid adapter before merging the URL into the config hash?

I'm also fine just closing this and having the adaptor monkey patch `DatabaseConfigurations`. I did rename it to `endpoint` for an rc release but been using `url` for 6+ years so want to support the `url` param for people using it, but if I'm going to support `url` anyways I don't see a need to rename it in the adapter.

Reply:
Your example would not only change the adapter but also the database. production would connect to the postgre adapter and the database foo.

I'm actually surprised you was able to use url for 6+ years. Where you able to bisect which commit change the behavior in 6.1?

Reply:
Yup! As of this commit ce9b197cc902ce8b72b7b3b5a3ec7e507fdba858.

Another option may be to generalize the treatment that happens with [JDBC](https://github.com/rails/rails/blob/main/activerecord/lib/active_record/database_configurations/url_config.rb#L45)?

Reply:
Was `url` introduced so people could put `url` in the YAML files? Or is it to patch in the ENV var?

Priority to the `ENV['URL']` makes sense if you use `ENV`, but seems messy in the YAML file since if can contain conflicting information especially with inheritance.

I'm not sure if we're able to read the inheritance of the YAML blocks, but if  we can we can use that to determine priority.

And in cases like:

```yaml
production:
  url: postgresql://host/foo
  adapter: sqlite3
  database: bar
```

We could throw the `ActiveRecord::DatabaseConfigurations::InvalidConfigurationError` error introduced by @eileencodes in dde03a9234e9b7fe802a45d511720c0ac3bf7617 if the adapter doesn't accept`:url` as a valid connection parameter.

Reply:
Yes. It was introduced to be used in the YAML files and it was introduced in Rails 4.1 if I remember correctly. https://github.com/rails/rails/pull/13541

Yes, it is true that url will have precedence over all conflicting configurations on the yaml and that is the intended behavior. This option, in my opinion, should not even be passed down to the adapter, since it is a Rails specific option

I'm really surprised that only in Rails 6.1 this behavior is now correct with this adapter.

In my opinion we should close the PR and the issue. It is unfortunate that things broke on Rails 6.1, but the behavior is not the correct and the adapter will need to adapt to this change that should really have happened in Rails 4.1.

Reply:
Interesting not sure how those test were passing, maybe it was just used to determine adapter and up to the adapter to discard url? But that doesn't make since because I've been loading the correct adapter not the "https" adapter until Rails 6.1.

Either way enough investigating for me I'm cool with the monkey patch, or maybe just alias as Http/Https Adapter 🤣.

Reply:
@rafaelfranca Or would making [`url.start_with?(...)`](https://github.com/rails/rails/blob/9fcaac6fb9a391046019a2399b6ede75a002e4c8/activerecord/lib/active_record/database_configurations/url_config.rb#L45) check against an array that adapter can append to be a solution?

Reply:
I think we can check against known schemas like `http`, `https` like we do with jdbc, but I think checking against know adapters will require the framework to change every time a new adapter is created and I want to avoid that.